[CSL-2184] Get rid of forks in NTP.Client

gromakovsky opened the pull request:
I slightly modified some logic to make it possible.
I think this code is still not perfect and there are other things
to improve, but within this issue I am only concerned about `forkIO` usages.
Notable changes (other than actual eliminating `forkIO`):
1. I removed stop button, because it's not needed anymore.
2. I removed `ncClosed`, because it seems to be not needed, because
 with current interface it seems to be impossible to access `NtpClient` after
 it's closed.
3. I moved `modifyLoggerName` to `spawnNtpClient` and it allowed to remove
ad-hoc `log` and `log'` functions.

Martoon-00 commented on networking/src/NTP/Client.hs:
```diff
@@ -280,22 +277,19 @@ stopNtpClient cli = do
+        withAsync (withSocketsDoLifted $ startReceive cli) $ \_ ->
+            withAsync (withSocketsDoLifted $ startSend addrs cli) $ \_ -> do
+                log cli Info "Launched NTP client"
+                callback (NtpStopButton $ stopNtpClient cli)
```
Since it's now written this way, it seems better to press stop button here (which will close sockets) rather than export button outside

Martoon-00 replied:
Also, this pretty code
https://github.com/input-output-hk/cardano-sl/blob/d9e2fd91bc9e1a7b8a3d933c55c5a5f37c32c6dd/networking/src/NTP/Client.hs#L256:L259
becomes unsafe because exception can fly in during `=<<` and created socket will remain unused :confused:

gromakovsky replied:
> Also, this pretty code becomes unsafe

Was it safe before? To be honest the whole design of `NTP.Client` looks not perfect, but in this PR I only wanted to spawn threads properly.

Martoon-00 commented on networking/src/NTP/Client.hs:
```diff
@@ -280,22 +277,19 @@ stopNtpClient cli = do
-
-        return $ NtpStopButton $ stopNtpClient cli
+        when (null addrs) $ throwM NoHostResolved
+        withAsync (withSocketsDoLifted $ startReceive cli) $ \_ ->
```
`withSocketsDo` uses global counter to track whether Windows socket library should be attached, so one `withSocketsDoLifted` in the beginning would be enough

gromakovsky replied:
> `withSocketsDo` uses global counter to track whether Windows socket library should be attached, so one `withSocketsDoLifted` in the beginning would be enough

As you can see, old code was using it more than once.
```
            void . fork . withSocketsDoLifted $ startReceive cli
            void . fork . withSocketsDoLifted $ startSend addrs cli
```

So I decided to preserve old behaviour. Do I understand correctly that in old code usage of these `withSocketsDoLifted` was redundant as well?

Martoon-00 replied:
In old code you could not know for sure (at least at first sight) which of the threads will finish earlier, and both require socket library attached, so both are wrapped into `withSocketsDo`.
Now you can happily use only single `withSocketsDo` which wraps over all threads.

Martoon-00 commented on networking/src/NTP/Client.hs:
```diff
@@ -1,47 +1,46 @@
+
+import           Control.Concurrent.STM (modifyTVar')
+import           Control.Concurrent.STM.TVar (TVar, readTVar, writeTVar)
+import           Control.Exception.Safe (Exception, MonadMask, bracketOnError, catchAny, handleAny)
```
I didn't notice how we switched to `exceptions-safe` here (which is indeed nice, also my comment isn't related to exactly this PR, but still), looks like migration was done not fully correctly. 

For instance, 
https://github.com/input-output-hk/cardano-sl/blob/d9e2fd91bc9e1a7b8a3d933c55c5a5f37c32c6dd/networking/src/NTP/Client.hs#L278
assumed to be cleanup action, and shouldn't be interrupted by async exceptions

gromakovsky commented:
@Martoon-00 please have a look at my last commit. I think all your concerns are now addressed. I went further simplified `NTP.Client` a bit more. Hopefully I didn't break anything, but please review it carefully.
Notable changes:
1. I removed stop button, because it's not needed anymore.
2. I removed `ncClosed`, because it seems to be not needed, because with current interface it seems to be impossible to access `NtpClient` after it's closed.
3. I moved `modifyLoggerName` to `spawnNtpClient` and it allowed to remove ad-hoc `log` and `log'` functions.
4. I ditched the idea of passing continuation to `withNtpClient`, now we have `spawnNtpClient`. The caller can use `withAsync spawnNtpClient` or use `race` or just run it in main thread, whatever.

>  forM_ sockets $ \s -> liftIO (close s) `catchAny` (const $ pure ()) 

This code doesn't exist anymore. Sockets are always closed as release action of `bracket`.

I also left only one `withSocketsDoLifted`.

gromakovsky commented:
If everything looks fine, I'll squash two commits and write a bit better commit message.

Martoon-00 commented:
> > Also, this pretty code becomes unsafe
> Was it safe before? To be honest the whole design of NTP.Client looks not perfect, but in this PR I only wanted to spawn threads properly.
Yeees, it was not perfect. Despite this fact, previously `startSend` and `startReceive` were assumed to stop on their own and didn't expect any async exceptions (which was wrong in presence of different `MemoryOverflow` and others, but still ~plausible). But now dropping `ThreadKilled` in these functions is normal case, so async-exceptions safety should be revised at new scale.

Martoon-00 commented:
All other things now look much better than they were, thanks for your big work!

gromakovsky commented:
> Yeees, it was not perfect. Despite this fact, previously startSend and startReceive were assumed to stop on their own and didn't expect any async exceptions (which was wrong in presence of different MemoryOverflow and others, but still ~plausible). But now dropping ThreadKilled in these functions is normal case, so async-exceptions safety should be revised at new scale.

Okay, I see. Anyway, I think that:
1. Probability of such situation is very small anyway, because first there should be an exception in `doReceive` and second it should be killed during `=<<` which is really unlikely to happen. I am not saying that we shouldn't care at all, but I think that probability is not much higher than the probability of `MemoryOverflow` and others.
2. I think it's an indicator of another flaw in the design on `NTP.Client` and in this PR there is no goal to address all such flaws. So I'd like to leave it for further improvements.

So, @Martoon-00, I squashed two commits, can you approve now or do you want me to do anything else here?

Martoon-00 commented:
No, everything looks good to me.
`<access to green button granted>`
